lib/stringToParts.js

```diff
@@ -1,46 +1,62 @@
 'use strict';
 
+const split = String.prototype.split;
+const numbers = new Set(['0', '1', '2', '3', '4', '5', '6', '7', '8', '9']);
+
 module.exports = function stringToParts(str) {
-  const result = [];
+  if (str.indexOf('[') === -1) {
+    if (str.indexOf('.') !== -1) {
+      return split.call(str, '.');
+    }
+    return [str];
+  }
+
+  const result = [],
+      len = str.length;
+
+  let curPropertyName = '',
+      i = 0,
+      state = 0,
+      char = '';
+
+  for (i = 0; i < len; ++i) {
+    char = str[i];
 
-  let curPropertyName = '';
-  let state = 'DEFAULT';
-  for (let i = 0; i < str.length; ++i) {
     // Fall back to treating as property name rather than bracket notation if
     // square brackets contains something other than a number.
-    if (state === 'IN_SQUARE_BRACKETS' && !/\d/.test(str[i]) && str[i] !== ']') {
-      state = 'DEFAULT';
+    if (state === 1 && char !== ']' && !numbers.has(char)) {
```

Comment:
Is using 1 instead of 'IN_SQUARE_BRACKETS' noticably faster? I'm not sure the readability tradeoff is worth the extra perf.

Reply:
using now constants instead of magic numbers